### ISSUE TITLE
NetworkPolicy tests for blocking north/south traffic

### ISSUE THREAD

**Issue by danwinship** [kubernetes/kubernetes#114369]
The NP docs point out that [NP's semantics for north/south traffic are not very clearly defined](https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-ipblock-selectors):

> Cluster ingress and egress mechanisms often require rewriting the source or destination IP of packets. In cases where this happens, it is not defined whether this happens before or after NetworkPolicy processing, and the behavior may be different for different combinations of network plugin, cloud provider, `Service` implementation, etc.
> 
> In the case of ingress, this means that in some cases you may be able to filter incoming packets based on the actual original source IP, while in other cases, the "source IP" that the NetworkPolicy acts on may be the IP of a `LoadBalancer` or of the Pod's node, etc.
> 
> For egress, this means that connections from pods to `Service` IPs that get rewritten to cluster-external IPs may or may not be subject to `ipBlock`-based policies.

However, what's _not_ ambiguous is that a pod which is fully isolated for ingress should not accept E/W _or_ N/S traffic. Regardless of how the ingress/LB/cloud handles the traffic, it should end up getting blocked. (Well, unless it gets masqueraded to the pod's node IP, because then it hits that exception to the rules.)

Unfortunately we can't say the same for egress; I think we assume that a pod-to-service-IP connection will be allowed to reach kube-proxy even if the pod is fully isolated-for-egress, but we explicitly don't require that `ipBlock` policies get applied after service proxying.

Anyway, we should be able to add a test to `test/e2e/network/netpol/network_policy.go` that confirms that cluster-ingress traffic to a fully isolated-for-ingress pod is not allowed. In particular, if we create a LoadBalancer Service with `externalTrafficPolicy: Local`, and a NetworkPolicy blocking all ingress to that Service's Pods, then we should not be able to connect to the service via either the LoadBalancer IP/name or via its NodePort (on one of the correct nodes).

/sig network
/area network-policy
/priority backlog
/help
/good-first-issue

**Comment by MohnishMishra** [comment]
Hello, I am Mohnish Mishra. I am a 2nd year student in BTech in Computer Science and Engineering. 
I am new to open-source coding and trying understand the same. 
I am interested in contributing to this project. 
If you could please guide me on what is to be done and what are the pre-requisites for the same, it will be of great help.

**Comment by sftim** [comment]
@danwinship this feels to me like an issue that (right now) justifies the help label, but maybe not good-first-issue. WDYT?

**Comment by astoycos** [comment]
This has been on my TODO list, and I'm hoping to either get to it early in the new year or have a member of the network policy api [subgroup](https://github.com/kubernetes/community/tree/master/sig-network) tackle this. @MohnishMishra Feel free to come to that meeting to learn more :) [Note we are changing the time this year see [this pr](https://github.com/kubernetes/community/pull/7020) and [slack conversation](https://kubernetes.slack.com/archives/C01CWSHQWQJ/p1671475245599189) for more information]

But yeah @sftim I'd say this is _slightly_ above good first issue :)

**Comment by farahtk** [comment]
Hey. If this is not being actively worked on, I would like to pick this up.

**Comment by MohnishMishra** [comment]
Thank you @astoycos. 
I have joined the mailing group and hope to join the meetings and learn. 
Thank you for the opportunity.

**Comment by aayush1205** [comment]
If there is no active development here, I'd like to take it up!

**Comment by nag-geek** [comment]
Hy, I'm Nag, im new to this, i'd like to contribute to this open source i don't know from where to start. could you help me guys.

**Comment by astoycos** [comment]
/assign @daman1807 

who has been very active in the KPNG community and it going to pick this up :) thanks!

**Comment by jayunit100** [comment]
@daman if you use metallb you can directly implement 9 services, for the 9 pods in the reachability matrix, and then poll them all using dans idea

**Comment by jayunit100** [comment]
looking at this test w/ daman, it seems like an interesting test, but it seems extremely dependent on the infrastructure provider ... i wonder if maybe we should define it in some way as a "networking integration" test, and run it on specific setups - i.e. maybe where the e2e thing is running outside of the cluster, and so on....  ?

**Comment by danwinship** [comment]
> looking at this test w/ daman, it seems like an interesting test, but it seems extremely dependent on the infrastructure provider

How so? If both LoadBalancers and NetworkPolicy are available, then it _must_ be the case that LoadBalancer traffic is blocked when there is a NetworkPolicy blocking all ingress. (Unless the LoadBalancer ingress path ends up SNATting traffic to the pod's node IP, hence the use of `externalTrafficPolicy: Local` to ensure there is no SNAT.)

**Comment by jayunit100** [comment]
If you try to access an external service of type LB from a node... the kube proxy will iirc actually not go through the LB and cheat going straight to the pod...

In our e2es since usually we run the test from a pod ... the kube proxy impl therefore I think maybe is the thing being tested just as much as the netpol impl

But I guess we could add these tests.  In general I guess I'd wonder what the "right" way to run them is: would users wanna run em outside the cluster to test the realistic data path that flows from service LB into the pool?

**Comment by danwinship** [comment]
The LoadBalancer e2e tests, in general, connect from the host running `e2e.test`, not from a host inside the cluster, since, as you note, otherwise we would not actually be testing the load balancers.

**Comment by ashutosh887** [comment]
let me work on this issue now please
/assign

**Comment by jayunit100** [comment]
Ok....this feels like a loadbalancer e2e more then a netpol one but either way works .... let's just make sure we log that it must be run from outside a pod to be meaningful

**Comment by arashgodgiven** [comment]
Hi all, I'm a bit new to this. It seems that to be able to contribute to kubernetes we must be a kubernetes member, and to become one, we must ask current kubernetes members to sponsor us? Am I correct?
If so, would anyone help me become a kubernetes member? I am a recent grad, computer science bachelor from SFU. I have 2 years of experience working as a software developer. I don't have much experience with kubernetes but I want to get started.
Thanks

**Comment by aojea** [comment]
> Hi all, I'm a bit new to this. It seems that to be able to contribute to kubernetes we must be a kubernetes member, and to become one, we must ask current kubernetes members to sponsor us? Am I correct? If so, would anyone help me become a kubernetes member? I am a recent grad, computer science bachelor from SFU. I have 2 years of experience working as a software developer. I don't have much experience with kubernetes but I want to get started. Thanks

@arashgodgiven please avoid commenting on existing issues :)

Start with sig-contribex, that is the group help new contributors https://github.com/kubernetes/community/blob/master/sig-contributor-experience/README.md

**Comment by danwinship** [comment]
> * **Test Location**

Probably `tests/e2e/network/netpol/network_policy.go` should be moved up a level to `test/e2e/network/network_policy.go` (based on the comments in that file, it seems like maybe that was the original plan?), and `netpol/` should only contain the helpers, not the actual e2e tests themselves. And then the new test can be added to `test/e2e/network/network_policy.go`?

Note that wherever it ends up, the test will need both `feature.NetworkPolicy` and `feature.LoadBalancer`.

> * **NodePort coverage**

The non-`externalTrafficPolicy: Local` case is expected to do SNAT and the `externalTrafficPolicy: Local` case for NodePort is not interestingly different from the LoadBalancer IP case, so there's probably no point in testing this separately.

> * **Underlying non-determinism**

It's not really "non-determinism"; any given environment is deterministic. And some people will run the e2e suite in environments that will fail the test due to SNAT. So I think the result should be:
- connection is blocked: **Pass**
- connection is accepted with original source IP preserved: **Fail**
- connection is accepted with source IP changed: **Skip**

I think having the test skip itself when it detects SNAT is preferable to just telling people "you should skip this test if your environment doesn't support it", because you're only supposed to skip it if the connection gets through because your load balancer did the "wrong" thing, _not_ if the connection gets through because your _network plugin_ does the wrong thing.

**Comment by justinlyon12** [comment]
Hey @Yash-xoxo, I've still been working on this and plan to push a PR this week!

**Comment by justinlyon12** [comment]
@danwinship, sorry for the slow progress on this one.  I wrote the new e2e test and moved the netpol tests one level up as you suggested in your above comment. 

To validate the  "connection is accepted with source IP changed: Skip" path however, I've spend quite a while (attempting) to reproduce, on KIND, the the scenario where SNAT lets traffic bypass the default-deny network policy (Using eTP: Cluster, as I'll touch on below).

As I was having difficultly I took a step back, and started to wonder if this branch is even needed for the following reasons:

- With eTP = Local, client IP preservation is mandatory (mentioned [here](https://github.com/kubernetes/enhancements/pull/4632#discussion_r1603346835)), so any SNAT would already violate the eTP = Local contract.
- The test, as I implemented it, is "should enforce default-deny ingress policy against LoadBalancer traffic with ExternalTrafficPolicy: Local". Based on this if the connection is accepted, then either:
  1. The NetPol was bypasssed
  2. eTP = Local didnt preserve the client IP (SNAT)

In both scenarios I would think the test should simply fail, rather than requiring special "SNAT detection" logic. I was wondering however if this matches what you had in mind or if I am misunderstanding something/ overlooking an edge case.

**Comment by danwinship** [comment]
> I've spend quite a while (attempting) to reproduce, on KIND, the the scenario where SNAT lets traffic bypass the default-deny network policy

oh... yeah, I think when I commented before I may have forgotten that we were only planning to test with eTP:Local. So yeah, probably you can just consider the test failing in that case.

**Comment by Yash-xoxo** [comment]
The emphasis on CNCF governance and community meetings highlights that Kubernetes is truly open source, not controlled by any single company, even though Google initiated it.
Is there a specific aspect of Kubernetes you'd like to explore further, like its architecture, how to deploy your first application, or how it compares to other orchestration tools?

- Just want to use it? Start with the kubernetes.io documentation
- Want to contribute or build from source? You'll need either Go or Docker installed, then clone the repo and run **make**

**Comment by kunalworldwide** [comment]
@danwinship before I open a new PR — @aroradaman raised whether this should go in `network-policy-api` instead of `test/e2e/network/netpol/network_policy.go`. This is a v1 NetworkPolicy test so my assumption is it stays in `netpol/`, but wanted to check before putting work into the wrong place.

Also want to confirm I'm reading your latest comments right: the test should create a pod with deny-all-ingress, expose it via NodePort with `externalTrafficPolicy: Local`, connect from the e2e host (not from inside a pod), and just fail if the connection succeeds. No SNAT detection needed. Should I include a LoadBalancer variant too or keep it narrow for now?

**Comment by danwinship** [comment]
@kunalworldwide there is already a PR open adding a test for this. It just hasn't merged yet.

**Comment by kunalworldwide** [comment]
@danwinship good to know — is the open PR #137390, or a different one? I saw #137390 was closed earlier; if there is a replacement PR it would be useful to link it here so others do not duplicate the work. Happy to help review or test if it is stuck on something.

**Comment by danwinship** [comment]
There have been multiple PRs opened at various points; you can see them all in the sidebar on the right. One of them is currently open.

**Comment by kunalworldwide** [comment]
@danwinship thanks — I see the open PR now. Is it stuck on anything specific (review bandwidth, rebase, or a design question)? Happy to help review or run the e2e tests if that unblocks it.
